mne/epochs.py

```diff
@@ -338,6 +339,26 @@ def _handle_event_repeated(events, event_id, event_repeated, selection,
     return new_events, event_id, selection, drop_log
 
 
+def _warn_empty(meth):
+    @wraps(meth)
+    def wrapper(*args, **kwargs):
+        # Prevent method from running if epochs are empty
+        if len(args[0]) == 0:
+            warn(f'{meth.__name__} can not run because Epochs are empty!')
+        else:
+            # Catch exceptions when epochs are dropped
+            # during execution of the method
+            try:
+                return meth(*args, **kwargs)
+            except (ZeroDivisionError, ValueError, IndexError):
```

Comment:
I dont like much exceptions that hide where exactly the problem is. I would rather be explicit in the wrapper. If it crashes for empty epochs rather than catching an exception we should just immediately raise one. It makes no sense to compute the psd of empty epochs for example.


Also can I ask you to add a tests calling all wrapped methods on some empty epochs?

🙏

Reply:
> It makes no sense to compute the psd of empty epochs for example.

The tricky part here is that things like `len(epochs)` / epoch count only exist once bads have been dropped, which sometimes isn't available until *after* running the method that iterates over the epochs to see if they're valid (i.e., with non-preloaded epochs).

Naively we could do `self.drop_bad()` and then check, but this is inefficient because it could cause iteration over epochs (reading from disk!) twice instead of once.

One safe way out of this I think might be to:

1. change the `__iter__` method to have a `raise_empty=False`, that these methods should set to True if no valid epochs are found
2. change the `_check_preload` function to have a `raise_empty=False` as well
3. ensure that any methods that need non-empty Epochs either iterate using `__iter__(..., raise_empty=True)` or call `_check_preload(..., raise_empty=True)

Something like this I think could work

Reply:
> Also can I ask you to add a tests calling all wrapped methods on some empty epochs?

Yes

>The tricky part here is that things like len(epochs) / epoch count only exist once bads have been dropped, which sometimes isn't available until after running the method that iterates over the epochs to see if they're valid (i.e., with non-preloaded epochs).

Yes, I think this is also why the tests are failing at the moment.

> 1. change the __iter__ method to have a raise_empty=False, that these methods should set to True if no valid epochs are found

@larsoner Just curious, would there be a way to pass `raise_empty=True` into an iteration call like `for epo in Epochs()` too/use `iter()`? Or does it have to be `__iter__` then?

Reply:
I think you have to use `__iter__` directly, which should be fine

Reply:
> 2. change the _check_preload function to have a raise_empty=False as well

I am not sure if this is necessary, since only `BaseEpochs.crop` and `BaseEpochs.apply_function` are using `_check_preload` and this helper-function is used by a wide range of object-types to have this specific signature change. WDYT?

Reply:
Maybe I understand something wrong here, but there seems to be a paradox for me:
__iter__ is at the beginning when we iterate the Epochs-object to check if it has any epochs, right?
How can we raise empty in this method then if we haven't iterated over it yet?

Reply:
@larsoner In which case do we have empty epochs without having called `Epochs.drop_bad()` internally or externally? (isn't this the most common reason for empty eochs?)

Reply:
@marsipu I think you miss the point a bit here about `epochs.drop_bad` / `epochs.drop`. I am talking about how any "empty" check needs to be implemented. There are two ways I know of that our methods that operate on epochs currently get the data they need:

1. They iterate over epochs and take the data from each and use it in some way via `for epoch in epochs` or so. Hence my suggestion that, for such functions, being able to do `for epoch in epochs.__iter__(raise_error=True)` might be useful for some of them and `raise_error=False` might be better in others, depending on whether or not a sensible value can be returned in the case of empty epochs.
2. `_check_preload` then use `_data` directly -- hence my point (2) above about modifying `_check_preload` to look for emptiness for functions where empty epochs would again lead to something that doesn't make sense..

And then point (3) is just a suggestion to use these two where needed.